### PR TITLE
feat: implement add_transaction

### PR DESCRIPTION
**Pull request by ayeletstarkware**
This change is [Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/6)

### CODE REVIEW

**Review by ayeletstarkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/6)* status: 0 of 8 files reviewed, 1 unresolved discussion (waiting on @ArniStarkware, @elintul, and @giladchase)

___
*[`crates/gateway/src/test_utils.rs` line 1 at r1](https://reviewable.io/reviews/starkware-libs/mempool/6#-NtBGlJU77JqyAWN-_pZ:-NtBGlJU77JqyAWN-_p_:b-orferh) ([raw file](https://github.com/starkware-libs/mempool/blob/dcfa712cd5dbb87365b708f2d0a7342513657942/crates/gateway/src/test_utils.rs#L1)):*
> ```XML
> use starknet_api::core::ClassHash;
> ```

Should this file be placed inside a folder named `test_utils` with `transaction.rs` as its name?

**Review by ayeletstarkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/6)* status: 0 of 8 files reviewed, 12 unresolved discussions (waiting on @ArniStarkware, @elintul, and @giladchase)

___
*[`crates/gateway/Cargo.toml` line 12 at r2](https://reviewable.io/reviews/starkware-libs/mempool/6#-NtLUnxMEXOHBpVozMXI:-NtMPJuT0pblojaxom7n:b-896fix) ([raw file](https://github.com/starkware-libs/mempool/blob/49b3dc9c0d2f59d120c260f4bb150c2f258cea17/crates/gateway/Cargo.toml#L12)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

Are we using this here? if not then remove, we'll put it back once it's used
</blockquote></details>

Done.
___
*[`crates/gateway/src/gateway_test.rs` line 58 at r2](https://reviewable.io/reviews/starkware-libs/mempool/6#-NtLa7B4Ern35zgEF0Oj:-NtPuDlZ2ZF4r0_GoB-O:b-896fix) ([raw file](https://github.com/starkware-libs/mempool/blob/49b3dc9c0d2f59d120c260f4bb150c2f258cea17/crates/gateway/src/gateway_test.rs#L58)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

you can unwrap in tests unless the `expect` is documenting some assumption that isn't immediately obvious from the code
</blockquote></details>

Done.
___
*[`crates/gateway/src/gateway_test.rs` line 81 at r2](https://reviewable.io/reviews/starkware-libs/mempool/6#-NtLaHYd3JqwAJiWGujm:-NtPu0Qr1G1rXZ5qG5Vc:b-896fix) ([raw file](https://github.com/starkware-libs/mempool/blob/49b3dc9c0d2f59d120c260f4bb150c2f258cea17/crates/gateway/src/gateway_test.rs#L81)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

use assert_matches!
</blockquote></details>

Done.

**Comment by ArniStarkware** [comment]
*[`crates/gateway/src/gateway.rs` line 65 at r2](https://reviewable.io/reviews/starkware-libs/mempool/6#-NtLVmSZ1MJ4C2aMz45U:-NtPih13Bh3xJ-7y1asx:brafrq7) ([raw file](https://github.com/starkware-libs/mempool/blob/49b3dc9c0d2f59d120c260f4bb150c2f258cea17/crates/gateway/src/gateway.rs#L65)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

Any way of making the arg be Transaction? that is, triggering the deserde automatically (similar to pyo3)?
(Can't believe I said "similar to pyo3" :see_no_evil: : )
</blockquote></details>

At some point, we need a function that converts `hyper::Body` to `Transaction`/`ExternalTransaction`. Or, was that exactly your point about `pyo3`?

What do you think about using a similar design to the main repo? With `Gateway` and `GatewayImpl`?
The `GatewayImpl` should indeed get `ExternalTransaction`.

**Review by giladchase — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/6)* status: 0 of 8 files reviewed, 8 unresolved discussions (waiting on @ArniStarkware, @ayeletstarkware, and @elintul)

___
*[`crates/gateway/src/gateway.rs` line 65 at r2](https://reviewable.io/reviews/starkware-libs/mempool/6#-NtLVmSZ1MJ4C2aMz45U:-NtQaeje1y8kUc5xjvES:b-k2nfvx) ([raw file](https://github.com/starkware-libs/mempool/blob/49b3dc9c0d2f59d120c260f4bb150c2f258cea17/crates/gateway/src/gateway.rs#L65)):*
<details><summary><i>Previously, ArniStarkware (Arnon Hod) wrote…</i></summary><blockquote>

At some point, we need a function that converts `hyper::Body` to `Transaction`/`ExternalTransaction`. Or, was that exactly your point about `pyo3`?

What do you think about using a similar design to the main repo? With `Gateway` and `GatewayImpl`?
The `GatewayImpl` should indeed get `ExternalTransaction`.
</blockquote></details>

Unblocking.

Lets talk about this after we TAL at `axum`, I think it can handle this.

Basically, I don't like the `Body` arg, which feels too close to an `Any`-typed arg.
I want type-checking in the signature.

**Review by giladchase — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/6)* status: 0 of 8 files reviewed, 9 unresolved discussions (waiting on @ArniStarkware, @ayeletstarkware, and @elintul)

___
*[`crates/gateway/src/gateway_test.rs` line 56 at r3](https://reviewable.io/reviews/starkware-libs/mempool/6#-NtQw-UL6vcjQGnfVCX5:-NtQw-UM65Pkbgcq08A6:br94s4a) ([raw file](https://github.com/starkware-libs/mempool/blob/a83cc4f155ece8c7a547ca8658839a865976ba70/crates/gateway/src/gateway_test.rs#L56)):*
> ```XML
> 
> #[tokio::test]
> async fn test_add_transaction_declare() {
> ```

Any other tx type plz :) like invoke.

Reason: supporting declare requires also sending a ContractClass, not just a Transaction.
No need to support this, I'm currently working on it rebased of this branch.



_Code quote:_
```XML
_declare
```

**Review by elintul — COMMENTED** [review]
> Reviewed 1 of 8 files at r1, 2 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/6)* status: 3 of 8 files reviewed, 10 unresolved discussions (waiting on @ArniStarkware and @ayeletstarkware)

___
*[`Cargo.toml` line 15 at r3](https://reviewable.io/reviews/starkware-libs/mempool/6#-NtkbMLP1NnFU8C25WDm:-NtkbMLP1NnFU8C25WDn:b-d6b84w) ([raw file](https://github.com/starkware-libs/mempool/blob/a83cc4f155ece8c7a547ca8658839a865976ba70/Cargo.toml#L15)):*
> ```TOML
> hyper = "0.13.9"
> tokio = { version = "0.2", features = ["full"] }
> thiserror = "1.0"
> ```

Please alphabetize, x2.

**Review by ayeletstarkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/6)* status: 0 of 15 files reviewed, 9 unresolved discussions (waiting on @ArniStarkware, @elintul, and @giladchase)

___
*[`crates/gateway/src/gateway.rs` line 65 at r2](https://reviewable.io/reviews/starkware-libs/mempool/6#-NtLVmSZ1MJ4C2aMz45U:-Nu-e4t9-zA-zoyxKEWI:b-a8k3rw) ([raw file](https://github.com/starkware-libs/mempool/blob/49b3dc9c0d2f59d120c260f4bb150c2f258cea17/crates/gateway/src/gateway.rs#L65)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

Unblocking.

Lets talk about this after we TAL at `axum`, I think it can handle this.

Basically, I don't like the `Body` arg, which feels too close to an `Any`-typed arg.
I want type-checking in the signature.
</blockquote></details>

Next PR.
___
*[`crates/gateway/src/gateway.rs` line 76 at r2](https://reviewable.io/reviews/starkware-libs/mempool/6#-NtLWJKr3Pn5XCxGEGT0:-Ntv0hqt8XBDwgfnHnpK:b-896fix) ([raw file](https://github.com/starkware-libs/mempool/blob/49b3dc9c0d2f59d120c260f4bb150c2f258cea17/crates/gateway/src/gateway.rs#L76)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

Leverage thiserror's `#from` attribute, possibly with `transparent`.

Then, either remove the map_err (because `?` will do it implicitly), or remove the Ok() and the `?`, leaving only the map_err.
</blockquote></details>

Done.
___
*[`crates/gateway/src/gateway_test.rs` line 56 at r3](https://reviewable.io/reviews/starkware-libs/mempool/6#-NtQw-UL6vcjQGnfVCX5:-Nu-e6XYEEohvbXtmgBw:bb6un1j) ([raw file](https://github.com/starkware-libs/mempool/blob/a83cc4f155ece8c7a547ca8658839a865976ba70/crates/gateway/src/gateway_test.rs#L56)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

Any other tx type plz :) like invoke.

Reason: supporting declare requires also sending a ContractClass, not just a Transaction.
No need to support this, I'm currently working on it rebased of this branch.
</blockquote></details>

Done
___
*[`crates/gateway/src/test_utils.rs` line 1 at r1](https://reviewable.io/reviews/starkware-libs/mempool/6#-NtBGlJU77JqyAWN-_pZ:-Nu-e8uDD73RnapIVelN:bhofej8) ([raw file](https://github.com/starkware-libs/mempool/blob/dcfa712cd5dbb87365b708f2d0a7342513657942/crates/gateway/src/test_utils.rs#L1)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

Should probably be in SN API's test_utils, since the blockifier probably has something similar.

Lets move it there, and if you have a spare moment PR the blockifier to use it, or alternatively add a Monday task to do it and let Noa know.
</blockquote></details>

Deleted
___
*[`crates/gateway/src/test_utils.rs` line 9 at r2](https://reviewable.io/reviews/starkware-libs/mempool/6#-NtLY5DU680m7XuPx9UZ:-Ntv2KSl9vZ5bHgZXwyL:bhofej8) ([raw file](https://github.com/starkware-libs/mempool/blob/49b3dc9c0d2f59d120c260f4bb150c2f258cea17/crates/gateway/src/test_utils.rs#L9)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

Fee::default(), but don't write it.
Remove this and the other default fields, and add `..Default::default()` at the end of the constructor.

Rationale: stating something as a "default value" communicates better how the reader should treat it, as opposed to: "reader readers Fee(0), then realizes that it's a default value", basically the reader mentally the user is transforming all of the explicit "default-valued" fields into a "the rest are defaults", so `..default::Default()` saves that extra mental step.
</blockquote></details>

Deleted
___
*[`crates/gateway/src/test_utils.rs` line 11 at r2](https://reviewable.io/reviews/starkware-libs/mempool/6#-NtLXwPe3JlSLZcJcTLl:-Nu-eC2u97kfU8CMAiam:bhofej8) ([raw file](https://github.com/starkware-libs/mempool/blob/49b3dc9c0d2f59d120c260f4bb150c2f258cea17/crates/gateway/src/test_utils.rs#L11)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

StarkFelt::ONE (its a const, better perf)
</blockquote></details>

Deleted
___
*[`crates/gateway/src/test_utils.rs` line 13 at r2](https://reviewable.io/reviews/starkware-libs/mempool/6#-NtL_xof69Nx8BVWxKBz:-Nu-eCt4CKHzGi7M1fux:bhofej8) ([raw file](https://github.com/starkware-libs/mempool/blob/49b3dc9c0d2f59d120c260f4bb150c2f258cea17/crates/gateway/src/test_utils.rs#L13)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

use `contract_address!`, it is part of SNAPI's testing feature
</blockquote></details>

Deleted
___
*[`crates/gateway/src/transaction.rs` line 7 at r2](https://reviewable.io/reviews/starkware-libs/mempool/6#-NtL_da3BBmEWlZlYmW4:-Ntv1AwP1weVth44fkFU:b7yntr9) ([raw file](https://github.com/starkware-libs/mempool/blob/49b3dc9c0d2f59d120c260f4bb150c2f258cea17/crates/gateway/src/transaction.rs#L7)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

Can this be public?

Private fields in general must have some invariant on them that we are protecting (like a field that is a hashmap that our code assumes that contain certain keys).
</blockquote></details>

using ExternalTransaction from Starknet API (was copied from papyrus). Declare tx contains contract class.
___
*[`crates/gateway/src/transaction.rs` line 23 at r2](https://reviewable.io/reviews/starkware-libs/mempool/6#-NtLZqi761oY7XOs9x0V:-NtW_hhcAV5sZC-clEfs:b3kpx) ([raw file](https://github.com/starkware-libs/mempool/blob/49b3dc9c0d2f59d120c260f4bb150c2f258cea17/crates/gateway/src/transaction.rs#L23)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

We already have some of this logic in the blockifier's `Transaction` struct, via it's internal `TransactionType` mechanism.

We should avoid duplicating logic, do we want to maybe extract that bit from the blockifier and into SNAPI?
</blockquote></details>





_Code snippet:_
```XML
impl AccountTransaction {
    pub fn tx_type(&self) -> TransactionType {
        match self {
            AccountTransaction::Declare(_) => TransactionType::Declare,
            AccountTransaction::DeployAccount(_) => TransactionType::DeployAccount,
            AccountTransaction::Invoke(_) => TransactionType::InvokeFunction,
        }
    }
}
```

**Review by ayeletstarkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/6)* status: 0 of 15 files reviewed, 1 unresolved discussion (waiting on @ArniStarkware, @elintul, and @giladchase)

___
*[`crates/gateway/src/errors.rs` line 7 at r2](https://reviewable.io/reviews/starkware-libs/mempool/6#-NtLVEAR5tOgXjOSd5xu:-Nu3HPAQ1bbN9bFktJtG:b-896fix) ([raw file](https://github.com/starkware-libs/mempool/blob/49b3dc9c0d2f59d120c260f4bb150c2f258cea17/crates/gateway/src/errors.rs#L7)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

In general, if the custom string doesn't add anything on top of the variant, better remove it altogether, the variant name communicates this already.
(Though in almost all cases more info can be added on top of the variant name...)
</blockquote></details>

Done.

**Review by giladchase — CHANGES_REQUESTED** [review]
RE the large json files, i'm suspect this might be hard to maintain/ 
Every time we change one of the fields in the structs we'll have to fix json files, which isn't fun.
Perhaps if the values in the jsons were significant it'll be worth it, but since they're random, it might be easier to use a `new_for_testing` and the serialize it into json (or use a `default` constructor for the transaction).

WDYT?

> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/6#-:-NuKB2GEDdQdTgaJDVL5:bm963ti)* status: 0 of 15 files reviewed, 2 unresolved discussions (waiting on @ArniStarkware, @ayeletstarkware, and @elintul)

___
*[`crates/gateway/src/gateway.rs` line 33 at r6](https://reviewable.io/reviews/starkware-libs/mempool/6#-NuIL5oP6qu6DluYEI9x:-NuIL5oQ02AvXEQeVZTj:b-fqsjnr) ([raw file](https://github.com/starkware-libs/mempool/blob/26b54fe4c7623d033f5be5f297c6d2f24d224439/crates/gateway/src/gateway.rs#L33)):*
> ```XML
> 
>         Server::bind(&addr).serve(make_service).await?;
>         Server::bind(&addr).serve(make_service).await?;
> ```

Is this intentional?



_Code quote:_
```XML
        Server::bind(&addr).serve(make_service).await?;
        Server::bind(&addr).serve(make_service).await?;
```
___
*[`crates/gateway/src/gateway_test.rs` line 20 at r6](https://reviewable.io/reviews/starkware-libs/mempool/6#-NuK9d1h75l-A-RmXH9m:-NuK9d1h75l-A-RmXH9n:bpmgdhg) ([raw file](https://github.com/starkware-libs/mempool/blob/26b54fe4c7623d033f5be5f297c6d2f24d224439/crates/gateway/src/gateway_test.rs#L20)):*
> ```XML
> }
> 
> #[tokio::test]
> ```

add `rstest` dev-dep and use parameterization here, to reduce duplication.

**Review by ayeletstarkware — COMMENTED** [review]
It has been decided to support only V3, so many of the JSON files were deleted. I will add a `default` to StarkNet API's ExternalTransaction and update the tests with 'new_for_testing' in the next PR. For now, I've added a TODO.

> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/6#-:-NuTDuOv3CHtEZxUxvca:b-ljp6h6)* status: 0 of 15 files reviewed, 2 unresolved discussions (waiting on @ArniStarkware, @elintul, and @giladchase)

___
*[`crates/gateway/src/gateway.rs` line 33 at r6](https://reviewable.io/reviews/starkware-libs/mempool/6#-NuIL5oP6qu6DluYEI9x:-NuNsQK15hmiK70wZX1M:bb710yj) ([raw file](https://github.com/starkware-libs/mempool/blob/26b54fe4c7623d033f5be5f297c6d2f24d224439/crates/gateway/src/gateway.rs#L33)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

Is this intentional?
</blockquote></details>

Nope
___
*[`crates/gateway/src/gateway_test.rs` line 20 at r6](https://reviewable.io/reviews/starkware-libs/mempool/6#-NuK9d1h75l-A-RmXH9m:-NuTUOzZBRigsBZ0aenE:b-896fix) ([raw file](https://github.com/starkware-libs/mempool/blob/26b54fe4c7623d033f5be5f297c6d2f24d224439/crates/gateway/src/gateway_test.rs#L20)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

add `rstest` dev-dep and use parameterization here, to reduce duplication.
</blockquote></details>

Done.

**Review by giladchase — APPROVED** [review]
:lgtm:

> Reviewed 4 of 15 files at r4, 1 of 1 files at r5, 1 of 2 files at r6, 9 of 9 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/6#-:-NuY1Gz8A5rq4wZqDUyX:bnfp4nl)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @ArniStarkware and @elintul)

**Review by giladchase — APPROVED** [review]
> Reviewed 3 of 3 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/6)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @ArniStarkware and @elintul)
